daemon/api_snaps.go

```diff
@@ -260,6 +261,44 @@ func (inst *snapInstruction) holdLevel() snapstate.HoldLevel {
 	}
 }
 
+func cleanSnapshotOptions(snapshotOpts map[string]*snap.SnapshotOptions) {
+	for name, options := range snapshotOpts {
+		if options.Unset() {
+			delete(snapshotOpts, name)
+		}
+	}
+}
+
+// cleanAndValidateSnapshotOptions cleans and validates the snapshot options.
```

Comment:
(nitpick) sorry for being difficult here and I'm happy to be ignored but I'm not a fan of mixing functions that have side effects and functions should not have them. It feels to me like this should just be called `validateSnapshotOptions` and the cleanSnapshotOptions would be done in a separate step outside of the function. Then most of the explaination of the function could move to `cleanSnapshotOptions()` where it's better suited IMO and validate is clearly only about validate. This of course means there is a test needed that ensures that cleanSnapshotOptions is needed and I outlined a diff below (even if this is not changed I think we want the test :)

So this would become something like (of course the function name now needs an update and some tests need updating):
```go
diff --git a/daemon/api_snaps.go b/daemon/api_snaps.go
index 5d1624eef2..dfca63ce04 100644
--- a/daemon/api_snaps.go
+++ b/daemon/api_snaps.go
@@ -286,7 +286,6 @@ func (inst *snapInstruction) cleanAndValidateSnapshotOptions() error {
        if inst.Action != "snapshot" {
                return fmt.Errorf("snapshot-options can only be specified for snapshot action")
        }
-       cleanSnapshotOptions(inst.SnapshotOptions)
        for name, options := range inst.SnapshotOptions {
                if !strutil.ListContains(inst.Snaps, name) {
                        return fmt.Errorf("cannot use snapshot-options for snap %q that is not listed in snaps", name)
@@ -359,6 +358,9 @@ func (inst *snapInstruction) validate() error {
        if err := inst.cleanAndValidateSnapshotOptions(); err != nil {
                return err
        }
+       if inst.Action == "snapshot" {
+               cleanSnapshotOptions(inst.SnapshotOptions)
+       }
 
        return inst.snapRevisionOptions.validate()
 }
```

Reply:
Thanks, implemented.